import java.util.Iterator;
import java.util.LinkedList;
import java.util.Scanner;

class Qnode {
    int vert;
    int vertkey;

    Qnode(int v, int key) {
	vert = v;
	vertkey = key;
    }
}

class Heapmin {
    private static int posinvalida = 0;
    int sizeMax,size;

    Qnode[] a;
    int[] pos_a;

    Heapmin(int vec[], int n) {
	a = new Qnode[n + 1];
	pos_a = new int[n + 1];
	sizeMax = n;
	size = n;
	for (int i = 1; i <= n; i++) {
	    a[i] = new Qnode(i,vec[i]);
	    pos_a[i] = i;
	}

	for (int i = n/2; i >= 1; i--)
	    heapify(i);
    }

    int extractMax() {
	int vertv = a[1].vert;
	swap(1,size);
	pos_a[vertv] = posinvalida;  // assinala vertv como removido
	size--;
	heapify(1);
	return vertv;
    }

    void increaseKey(int vertv, int newkey) {

	int i = pos_a[vertv];
	a[i].vertkey = newkey;

	while (i > 1 && compare(i, parent(i)) < 0) {
	    swap(i, parent(i));
	    i = parent(i);
	}
    }

	//DeadCodeStart
    void insert(int vertv, int key)
    {
	if (sizeMax == size)
	    new Error("Heap is full\n");

	size++;
	a[size].vert = vertv;
	pos_a[vertv] = size;   // supondo 1 <= vertv <= n
	increaseKey(vertv,key);   // diminui a chave e corrige posicao se necessario
    }

    void write_heap(){
	System.out.printf("Max size: %d\n",sizeMax);
	System.out.printf("Current size: %d\n",size);
	System.out.printf("(Vert,Key)\n---------\n");
	for(int i=1; i <= size; i++)
	    System.out.printf("(%d,%d)\n",a[i].vert,a[i].vertkey);

	System.out.printf("-------\n(Vert,PosVert)\n---------\n");

	for(int i=1; i <= sizeMax; i++)
	    if (pos_valida(pos_a[i]))
		System.out.printf("(%d,%d)\n",i,pos_a[i]);
    }
	//DeadCodeEnd

    private int parent(int i){
	return i/2;
    }
    private int left(int i){
	return 2*i;
    }
    private int right(int i){
	return 2*i+1;
    }

    private int compare(int i, int j) {
	if (a[i].vertkey > a[j].vertkey)
	    return -1;
	if (a[i].vertkey == a[j].vertkey)
	    return 0;
	return 1;
    }


    private void heapify(int i) {
	int l, r, smallest;

	l = left(i);
	if (l > size) l = i;

	r = right(i);
	if (r > size) r = i;

	smallest = i;
	if (compare(l,smallest) < 0)
	    smallest = l;
	if (compare(r,smallest) < 0)
	    smallest = r;

	if (i != smallest) {
	    swap(i, smallest);
	    heapify(smallest);
	}

    }

    private void swap(int i, int j) {
	Qnode aux;
	pos_a[a[i].vert] = j;
	pos_a[a[j].vert] = i;
	aux = a[i];
	a[i] = a[j];
	a[j] = aux;
    }

	//DeadCodeStart
    private boolean pos_valida(int i) {
	return (i >= 1 && i <= size);
    }
	//DeadCodeEnd
}


class Arco {
    int no_final;
    int valor;

    Arco(int fim, int v){
	no_final = fim;
	valor = v;
    }

	//DeadCodeStart
    int extremo_final() {
	return no_final;
    }

    int valor_arco() {
	return valor;
    }
	//DeadCodeEnd
}


class No {
    //int label;
    LinkedList<Arco> adjs;

    No() {
	adjs = new LinkedList<Arco>();
    }
}


class Grafo {
    No verts[];
    int nvs, narcos;

    public Grafo(int n) {
	nvs = n;
	narcos = 0;
	verts  = new No[n+1];
	for (int i = 0 ; i <= n ; i++)
	    verts[i] = new No();
        // para vertices numerados de 1 a n (posicao 0 nao vai ser usada)
    }

	//DeadCodeStart
    public int num_vertices(){
	return nvs;
    }

    public int num_arcos(){
	return narcos;
    }

    public LinkedList<Arco> adjs_no(int i) {
	return verts[i].adjs;
    }
	//DeadCodeEnd

    public void insert_new_arc(int i, int j, int valor_ij){
	verts[i].adjs.addFirst(new Arco(j,valor_ij));
        narcos++;
    }

	//DeadCodeStart
    public Arco find_arc(int i, int j){
	for (Arco adj: adjs_no(i))
	    if (adj.extremo_final() == j) return adj;
	return null;
    }
	//DeadCodeEnd

    public int prim(){
    	int custos[] = new int[nvs+1];
    	boolean visitado[] = new boolean[nvs+1];
    	int pai[]= new int[nvs+1];
    	for (int x=1; x<=nvs; x++){
    		custos[x]=Integer.MIN_VALUE;
    		visitado[x]=false;
    	}
    	custos[1]=0;
    	Heapmin hm= new Heapmin(custos, nvs);
    	while (hm.size!=0){
    		int no= hm.extractMax();
    		if (custos[no]==Integer.MIN_VALUE){
    			return Integer.MIN_VALUE;
    		}
    		Iterator<Arco> adj= verts[no].adjs.iterator();
    		while(adj.hasNext()){
    			Arco ex= adj.next();
    			if (visitado[ex.no_final]==false && custos[ex.no_final]<ex.valor){
    				custos[ex.no_final]=ex.valor;
    				pai[ex.no_final]=no;
    				hm.increaseKey(ex.no_final, custos[ex.no_final]);
    			}

    		}
    		visitado[no]=true;
    		}
    	int t=0;
		for (int x=1; x<=nvs;x++){
			t+=custos[x];

    	}
    	return t;
    }

}

public class Minimalista {

	public static void main(String[] args) {
		// TODO Auto-generated method stub
		Scanner stdin= new Scanner(System.in);
		int nos = stdin.nextInt() , ramos=stdin.nextInt() , custo= stdin.nextInt();
		Grafo exemplo= new Grafo(nos);
		for (int x=0; x<ramos; x++){
			int in=stdin.nextInt(), out=stdin.nextInt() , valor= stdin.nextInt(); valor= valor-custo;
			exemplo.insert_new_arc(in, out, valor);
			exemplo.insert_new_arc(out, in, valor);
		}
		int b= exemplo.prim();
		if (b==Integer.MIN_VALUE){
			System.out.println("impossivel");
		}
		else{
			System.out.println("rendimento optimo: "+b);
		}


	}

}
